doc/kb_sdk_impl_methods.md

```diff
@@ -0,0 +1,201 @@
+        report += 'Avg Length: '+str(sum(lengths)/float(len(lengths))) + ' bp.\n'
+
+        # compute a simple contig length distribution
+        bins = 10
+        counts, edges = np.histogram(lengths, bins)
+        report += 'Contig Length Distribution (# of contigs -- min to max basepairs):\n'
+        for c in range(bins):
+            report += '   '+str(counts[c]) + '\t--\t' + str(edges[c]) + ' to ' + str(edges[c+1]) + ' bp\n'
+
+        reportObj = {
+            'objects_created':[{'ref':params['workspace_name']+'/'+params['output_contigset_name'], 'description':'Assembled contigs'}],
+            'text_message':report
+        }
+
+        reportName = 'megahit_report_'+str(uuid.uuid4())   # uuid4() instead of getnode() to ensure uniqueness
+        report_obj_info = ws.save_objects({
+                'id':info[6],
+                'objects':[
+                    {
+                        'type':'KBaseReport.Report',
+                        'data':reportObj,
+                        'name':reportName,
+                        'meta':{},
+                        'hidden':1,
+                        'provenance':provenance
+                    }
+                ]
+            })[0]
+
+        output = { 'report_name': reportName, 'report_ref': str(report_obj_info[6]) + '/' + str(report_obj_info[0]) + '/' + str(report_obj_info[4]) }
+```
+
+[\[Back to Edit Impl\]](#impl)
+
+##### <A NAME="impl-shell-tool"></A>5.6. Invoking Shell Tool
+
+```python
+        command_line_tool_params_str = " ".join(command_line_tool_params)
+        command_line_tool_cmd_str = " ".join(command_line_tool_path, command_line_tool_params_str)
+        
+        # run <command_line_tool>, capture output as it happens
+        self.log(console, 'running <command_line_tool>:')
+        self.log(console, '    '+command_line_tool_cmd_str))
+        p = subprocess.Popen(command_line_tool_cmd_str,
+                    cwd = self.scratch,
+                    stdout = subprocess.PIPE, 
+                    stderr = subprocess.STDOUT, shell = False)
+
+        while True:
+            line = p.stdout.readline()
+            if not line: break
+            self.log(console, line.replace('\n', ''))
+
+        p.stdout.close()
+        p.wait()
+        self.log(console, 'return code: ' + str(p.returncode))
+        if p.returncode != 0:
+            raise ValueError('Error running <command_line_tool>, return code: '+str(p.returncode) + 
+                '\n\n'+ '\n'.join(console))
+ ```
+
+[\[Back to Edit Impl\]](#impl)
+
+##### <A NAME="impl-adding-data"></A>5.7. Adding Data To Your Method
+
+Data that is supported by [KBase Data Types](doc/kb_sdk_data_types_table.md) should be added as a workspace object.  Other data that is used to configure a method may be added to the repo with the code.  Large data sets that exceed a reasonable limit (> 1 GB) should be added to a shared mount point.  This can be accomplished by contacting kbase administrators at http://kbase.us.
```

Comment:
I'm not sure what this means - does it mean that if you need some data to run your method every time you should stick it in the workspace? If it does, generally speaking, no you shouldn't. If it's small, put it in the repo. If it's large use reference data.